### PR TITLE
Flag parsing

### PR DESCRIPTION
**Pull request by astradzhao**
flags are now able to be parsed regardless of order & type (e.g. 'semantifly add --source-type local_file subcommand.go --index-path=/workspace/semantifly main.go -copy=false' is parsed correctly, and adds both subcommand.go and main.go to the index file)

### CODE REVIEW

File: src/subcommand.go [hunk 1]
```diff
@@ -15,6 +15,37 @@ func printCmdErr(e string) {
 	fmt.Println(fmt.Sprintf("%s\n Try --help to list subcommands and options.", e))
 }
 
+func parseArgs(args []string) map[string][]string {
```
Comment (fredxfred) [inline src/subcommand.go]: Please add a unit test for this!

Comment (astradzhao) [reply src/subcommand.go]: done

File: src/utils/utils.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package utils
+
+import "strings"
+
+func ParseArgs(args []string) map[string][]string {
```
Comment (fredxfred) [inline src/utils/utils.go]: I don't think we'll ever call this form outside subcommand.go so we should leave it there IMO. Having a utils file like this is generally considered something best avoided. If there is truly a stateless function that is called from a bunch of different files a utils file would be ok, but it's preferable in cases like these to put ParseArgs in the file where it's used.

Also, once you move it back to subcommand.go, you should make it unexported ```func parseArgs``` since we don't expect other classes to call it.

File: src/tests/utils/parse_args_test.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
+package tests
```
Comment (fredxfred) [inline src/tests/utils/parse_args_test.go]: once you unexport parseArgs you'll need to make this the same as subcommand's package to call the package-local parseArgs function

File: src/tests/utils/parse_args_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+import (
+	"reflect"
+	"testing"
+
+	"accretional.com/semantifly/utils"
+)
+
+func TestParseArgs(t *testing.T) {
```
Comment (fredxfred) [inline src/tests/utils/parse_args_test.go]: Can you add another test case for invalid input and other edge cases?

1. Empty args
2. [--flag1, --flag2]
3. [arg1, --flag1]

File: src/tests/utils/parse_args_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
+package tests
+
```
Comment (fredxfred) [inline src/tests/utils/parse_args_test.go]: sorry just remembered, if we're testing code in src/foo.go we want its unit tests to be in eg src/foo_test.go so we can use the same package and thus test unexported functions

File: src/utils/utils.go (hunk 2 above)
Comment (fredxfred) [inline src/utils/utils.go]: Let's not return a map because we're actually only expecting two lists (either of which could maybe be empty)

File: src/tests/utils/parse_args_test.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		},
+		{
+			name: "Complicated test: different types of flags and order of arguments",
+			args: []string{"--flag1", "value1", "arg1", "arg2", "-flag2=value2", "arg3", "-flag3", "value3"},
+			expected: map[string][]string{
+				"flags":    {"--flag1", "value1", "-flag2", "value2", "-flag3", "value3"},
+				"nonFlags": {"arg1", "arg2", "arg3"},
+			},
+		},
+	}
```
Comment (fredxfred) [inline src/tests/utils/parse_args_test.go]: We can add a triple - test case ```---```

File: src/tests/utils/parse_args_test.go (hunk 5 above)
Comment (astradzhao) [reply src/tests/utils/parse_args_test.go]: done

File: src/tests/utils/parse_args_test.go (hunk 4 above)
Comment (astradzhao) [reply src/tests/utils/parse_args_test.go]: done

File: src/tests/utils/parse_args_test.go (hunk 3 above)
Comment (astradzhao) [reply src/tests/utils/parse_args_test.go]: done

File: src/utils/utils.go (hunk 2 above)
Comment (astradzhao) [reply src/utils/utils.go]: done

File: src/tests/utils/parse_args_test.go (hunk 6 above)
Comment (astradzhao) [reply src/tests/utils/parse_args_test.go]: done

**Comment by fredxfred** [comment]
Nice!
